File: backend/app/schemas/targets.py
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field, field_validator


class TargetMetricCatalogItem(BaseModel):
    metric_key: str
    label: str
    unit: str


class TargetDomainCatalogItem(BaseModel):
    domain_key: str
    domain_label: str
    metrics: list[TargetMetricCatalogItem]


class TargetMetricCatalogResponse(BaseModel):
    items: list[TargetDomainCatalogItem]


class ClubTargetUpsertRequest(BaseModel):
    domain_key: str = Field(min_length=1, max_length=64)
    metric_key: str = Field(min_length=1, max_length=64)
    period_key: str = Field(min_length=1, max_length=32)
    period_start: date
    period_end: date
    target_value: float = Field(gt=0)

    @field_validator("domain_key", "metric_key", "period_key")
    @classmethod
    def _normalize_key(cls, value: str) -> str:
        return value.strip().lower()


class ClubTargetResponse(BaseModel):
    id: uuid.UUID
    club_id: uuid.UUID
    domain_key: str
    domain_label: str
    metric_key: str
    metric_label: str
    unit: str
    period_key: str
    period_start: date
    period_end: date
    target_value: float
    archived: bool
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ClubTargetListResponse(BaseModel):
    items: list[ClubTargetResponse]
    total_count: int

File: frontend/src/types/targets.ts
export interface TargetMetricCatalogItem {
  metric_key: string;
  label: string;
  unit: string;
}

export interface TargetDomainCatalogItem {
  domain_key: string;
  domain_label: string;
  metrics: TargetMetricCatalogItem[];
}

export interface TargetMetricCatalogResponse {
  items: TargetDomainCatalogItem[];
}

export interface ClubTarget {
  id: string;
  club_id: string;
  domain_key: string;
  domain_label: string;
  metric_key: string;
  metric_label: string;
  unit: string;
  period_key: string;
  period_start: string;
  period_end: string;
  target_value: number;
  archived: boolean;
  archived_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ClubTargetListResponse {
  items: ClubTarget[];
  total_count: number;
}

export interface ClubTargetUpsertInput {
  domain_key: string;
  metric_key: string;
  period_key: string;
  period_start: string;
  period_end: string;
  target_value: number;
}
